src/pretix/api/views/checkin.py

```diff
@@ -188,6 +193,254 @@ def check_rules_qs(self, queryset, name, value):
+            file__isnull=False,
+            pk=data[len("file:"):],
+        )
+    except (BaseValidationError, BaseBaseValidationError, IndexError):  # invalid uuid
+        raise BaseValidationError('The submitted file ID "{fid}" was not found.'.format(fid=data))
+    except CachedFile.DoesNotExist:
+        raise BaseValidationError('The submitted file ID "{fid}" was not found.'.format(fid=data))
+
+    allowed_types = (
+        'image/png', 'image/jpeg', 'image/gif', 'application/pdf'
+    )
+    if cf.type not in allowed_types:
+        raise BaseValidationError('The submitted file "{fid}" has a file type that is not allowed in this field.'.format(fid=data))
+    if cf.file.size > settings.FILE_UPLOAD_MAX_SIZE_OTHER:
+        raise BaseValidationError('The submitted file "{fid}" is too large to be used in this field.'.format(fid=data))
+
+    return cf.file
+
+def _redeem_process(*, checkinlists, raw_barcode, answers_data, datetime, force, checkin_type, ignore_unpaid, nonce,
+                    untrusted_input, user, auth, expand, pdf_data, request, questions_supported, canceled_supported,
+                    source_type='barcode', legacy_url_support=False, simulate=False, gate=None):
+    if not checkinlists:
+        raise BaseValidationError('No check-in list passed.')
+
+    list_by_event = {checkinlist.event_id: checkinlist for checkinlist in checkinlists}
+    prefetch_related_objects([checkinlist for checkinlist in checkinlists if not checkinlist.all_products], 'limit_products') # prefetch_related
+
+    device = auth if isinstance(auth, Device) else None
+    gate = gate or (auth.gate if isinstance(auth, Device) else None)
+
+    context = {
+            'request': request,
+            'expand': expand,
+    }
+
+    def _make_context(context, event):
+        return{
+            **context,
+            'event': event,
+            'pdf_data': pdf_data and (
+                user if user and user.is_authenticated else auth
+            ).has_event_permission(request.organizer, event, 'can_view_orders', request),
+        }
+
+    common_checkin_args = dict(
+        raw_barcode=raw_barcode,
+        raw_source_type=source_type,
+        type=checkin_type,
+        list=checkinlists[0],
+        datetime=datetime,
+        device=device,
+        gate=gate,
+        nonce=nonce,
+        forced=force,
+    )
+    raw_barcode_for_checkin = None
+    from_revoked_secret = False
+    if simulate:
+        common_checkin_args['__fake_arg_to_prevent_this_from_being_saved'] = True
+
+    queryset = _checkin_list_position_queryset(checkinlists, pdf_data=pdf_data, ignore_status=True, ignore_products=True).order_by(
+        F('addon_to').asc(nulls_first=True)
+    )
+
+
+
+    q = Q(secret=raw_barcode)
+
+    if raw_barcode.isnumeric() and not untrusted_input and legacy_url_support:
+        q |= Q(pk=raw_barcode)
+
+
+    op_candidates = list(queryset.filter(q)) #filtering queryset with q to fetch orderposition
```

Comment:
Don't know why you have to load many rows and then, use only the first item

```py
op = op_candidates[0]
```

You should use `first` or `earliest` method on queryset.

Reply:
I have added a case to check ambiguity if multiple are found